File: test-accesstoken-pkg/py/test_databoard.py
"""Exhaustive boards (databoard) resource verification — pulls imbrace from PyPI.

Auth: Access Token.

Covers:
  - Board CRUD: list, get, create, update, delete, export_csv
  - Fields:     create_field, update_field, delete_field
  - Items:      list_items, get_item, create_item, update_item, delete_item, search
  - Segments:   list_segments, create_segment, update_segment, delete_segment
  - Folders:    search_folders, create_folder, get_folder, update_folder,
                delete_folders, get_folder_contents, search_files
  - Skipped:    import_csv/excel, upload_*, generate_ai_tags (need binary fixtures — covered in test_files.py)

Lifecycle test creates real resources then deletes them.
"""
from __future__ import annotations
import json
import os
import sys
import time
from typing import Any

from dotenv import load_dotenv

load_dotenv()

from imbrace import ImbraceClient

ACCESS_TOKEN = os.environ.get("IMBRACE_ACCESS_TOKEN")
ORG_ID       = os.environ.get("IMBRACE_ORGANIZATION_ID")
GATEWAY      = os.environ.get("IMBRACE_GATEWAY_URL", "https://app-gatewayv2.imbrace.co")

if not ACCESS_TOKEN or not ORG_ID:
    print("Missing IMBRACE_ACCESS_TOKEN or IMBRACE_ORGANIZATION_ID")
    sys.exit(1)

client = ImbraceClient(access_token=ACCESS_TOKEN, organization_id=ORG_ID, gateway=GATEWAY, timeout=30)
b = client.boards

passed = 0
failed = 0
skipped_n = 0
failures: list[str] = []


def step(label: str, fn) -> None:
    global passed, failed
    sys.stdout.write(f"  - {label} ... ")
    sys.stdout.flush()
    try:
        t0 = time.time()
        result = fn()
        dt = int((time.time() - t0) * 1000)
        summary = json.dumps(result, default=str)[:120] if result is not None else ""
        print(f"OK [{dt}ms] {summary}")
        passed += 1
    except Exception as e:
        detail = str(e)[:300]
        print(f"FAIL {detail}")
        failed += 1
        failures.append(f"{label} -> {detail}")


def skip(label: str, reason: str) -> None:
    global skipped_n
    print(f"  - {label}  SKIP: {reason}")
    skipped_n += 1


def section(title: str) -> None:
    print(f"\n== {title} ==")


print(f"\n=== boards (databoard) - auth: ACCESS TOKEN (PyPI imbrace) ===")
print(f"gateway={GATEWAY}\norg={ORG_ID}")

# ── 1. Read-only list ───────────────────────────────────────────────────────
section("Read-only list")
first_board_id: str | None = None

def _list_boards():
    global first_board_id
    r = b.list(limit=5)
    data = r.get("data") if isinstance(r, dict) else r
    if data:
        first_board_id = data[0].get("_id") or data[0].get("id")
    return {"count": len(data) if data else 0, "first": first_board_id}
step("list (limit 5)", _list_boards)

if first_board_id:
    step("get(first_board_id)",        lambda: {"id": (b.get(first_board_id) or {}).get("_id")})
    step("list_items(first, limit=2)", lambda: b.list_items(first_board_id, limit=2))
    step("list_segments(first)",       lambda: b.list_segments(first_board_id))
    step("export_csv(first)",          lambda: {"length": len(b.export_csv(first_board_id) or "")})
else:
    skip("get / list_items / list_segments / export_csv", "no boards in org")

# ── 2. Board lifecycle ──────────────────────────────────────────────────────
section("Board lifecycle (create -> update -> delete)")
stamp = int(time.time() * 1000)
test_board_name = f"sdk-test-board-{stamp}"
created_board_id: str | None = None

def _create():
    global created_board_id
    r = b.create(name=test_board_name, description="SDK test board", type="General", show_id=False)
    created_board_id = r.get("_id") or r.get("id")
    return {"id": created_board_id, "name": r.get("name"), "type": r.get("type")}
step("create (type=General)", _create)

if created_board_id:
    step("update (rename)",
         lambda: {"id": (b.update(created_board_id, {"name": f"{test_board_name}-updated", "description": "updated"}) or {}).get("_id")})

    # ── 3. Fields ─────────────────────────────────────────────────────────────
    # createField returns Board not BoardField — extract last field from board.fields[]
    section("Fields lifecycle")
    created_field_id: str | None = None

    def _create_field():
        global created_field_id
        # SDK now returns BoardField directly (was Board pre-fix #6).
        r = b.create_field(created_board_id, {"name": "test_field", "type": "ShortText"})
        created_field_id = r.get("_id") or r.get("id")
        return {"fieldId": created_field_id, "fieldName": r.get("name"), "fieldType": r.get("type")}
    step("create_field (ShortText)", _create_field)

    if created_field_id:
        step("update_field (rename)",
             lambda: b.update_field(created_board_id, created_field_id, {"name": "test_field_renamed"}))
        step("delete_field",
             lambda: b.delete_field(created_board_id, created_field_id) or {"deleted": created_field_id})
    else:
        skip("update_field / delete_field", "create_field did not return a field id")

    # ── 4. Items ──────────────────────────────────────────────────────────────
    # createItem body: {fields: [{board_field_id, value}]}
    # updateItem body: {fields: [], data: [{key, value}]}  — asymmetric, SDK drift
    section("Items lifecycle")
    identifier_field_id: str | None = None

    def _get_id_field():
        global identifier_field_id
        board = b.get(created_board_id)
        for f in (board.get("fields") or []):
            if f.get("is_identifier"):
                identifier_field_id = f.get("_id")
                return {"identifierFieldId": identifier_field_id, "name": f.get("name")}
        return {"identifierFieldId": None}
    step("get board identifier field id", _get_id_field)

    created_item_id: str | None = None
    if identifier_field_id:
        def _create_item():
            global created_item_id
            r = b.create_item(created_board_id,
                              {"fields": [{"board_field_id": identifier_field_id, "value": "hello"}]})
            created_item_id = r.get("_id") or r.get("id")
            return {"id": created_item_id}
        step("create_item", _create_item)
    else:
        skip("create_item", "no identifier field on board")

    if created_item_id:
        step("get_item",
             lambda: {"id": (b.get_item(created_board_id, created_item_id) or {}).get("_id")})
        step("update_item [asymmetric body: fields[] + data[] — SDK drift]",
             lambda: b.update_item(created_board_id, created_item_id,
                                   {"fields": [], "data": [{"key": identifier_field_id, "value": "updated"}]}))
        step("list_items(limit=5)", lambda: b.list_items(created_board_id, limit=5))
        step("search(q='hello')",   lambda: b.search(created_board_id, q="hello", limit=5))
        step("delete_item",
             lambda: b.delete_item(created_board_id, created_item_id) or {"deleted": created_item_id})
    else:
        skip("get_item / update_item / list_items / search / delete_item", "create_item did not return an id")

    # ── 5. Segments ───────────────────────────────────────────────────────────
    section("Segments lifecycle")
    created_segment_id: str | None = None
    def _create_seg():
        global created_segment_id
        r = b.create_segment(created_board_id, {"name": f"sdk-test-segment-{stamp}", "filter": {}})
        created_segment_id = r.get("_id") or r.get("id")
        return {"id": created_segment_id, "name": r.get("name")}
    step("create_segment", _create_seg)

    if created_segment_id:
        step("update_segment (rename)",
             lambda: b.update_segment(created_board_id, created_segment_id,
                                      {"name": f"sdk-test-segment-{stamp}-updated"}))
        step("list_segments",
             lambda: {"count": len(b.list_segments(created_board_id) or [])})
        step("delete_segment",
             lambda: b.delete_segment(created_board_id, created_segment_id) or {"deleted": created_segment_id})
    else:
        skip("update_segment / delete_segment", "create_segment did not return an id")

    section("Cleanup board")
    step("delete(board)",
         lambda: b.delete(created_board_id) or {"deleted": created_board_id})
else:
    skip("update / fields / items / segments / delete", "create did not return an id")

# ── 6. KnowledgeHub Folders ────────────────────────────────────────────────
# SDK should default source_type="upload" + parent_folder_id="root"
section("KnowledgeHub folders lifecycle")
step("search_folders", lambda: {"count": len(b.search_folders() or [])})

created_folder_id: str | None = None
def _create_folder():
    global created_folder_id
    r = b.create_folder({
        "name": f"sdk-test-folder-{stamp}",
        "organization_id": ORG_ID,
        "source_type": "upload",
        "parent_folder_id": "root",
    })
    created_folder_id = r.get("_id") or r.get("id")
    return {"id": created_folder_id, "name": r.get("name")}
step("create_folder [SDK should default source_type+parent_folder_id]", _create_folder)

if created_folder_id:
    step("get_folder",            lambda: {"id": (b.get_folder(created_folder_id) or {}).get("_id")})
    step("get_folder(recursive)", lambda: b.get_folder(created_folder_id, recursive=True))
    step("get_folder_contents",   lambda: b.get_folder_contents(created_folder_id))
    step("search_files",          lambda: {"count": len(b.search_files(created_folder_id) or [])})
    step("update_folder (rename)",
         lambda: b.update_folder(created_folder_id, {"name": f"sdk-test-folder-{stamp}-updated"}))
    step("delete_folders",
         lambda: b.delete_folders([created_folder_id]))
else:
    skip("get_folder / update_folder / delete_folders", "create_folder did not return an id")

# ── 7. Skipped (need binary fixtures) ───────────────────────────────────────
section("Skipped (need binary fixtures)")
skip("import_csv / import_excel", "needs CSV/Excel file fixture")
skip("upload_board_file",          "needs file fixture")
skip("upload_file (folder)",       "needs file fixture — covered in test_files.py")
skip("generate_ai_tags",           "needs file fixture")
skip("get_link_preview",           "needs URL fixture")
skip("get_related_items / link_items / unlink_items", "needs related board")

print(f"\n=== Summary ===")
print(f"pass={passed}  fail={failed}  skip={skipped_n}")
if failures:
    print("Failures:")
    for f in failures:
        print(f"  - {f}")
sys.exit(1 if failed > 0 else 0)

File: test-accesstoken-pkg/ts/test-databoard.ts
/**
 * Exhaustive boards (databoard) resource verification — pulls @imbrace/sdk from npm.
 * Auth: Access Token.
 *
 * Covers:
 *   - Board CRUD: list, get, create, update, delete, reorder, exportCsv
 *   - Fields:     createField, updateField, deleteField, reorderFields, bulkUpdateFields
 *   - Items:      listItems, getItem, createItem, updateItem, deleteItem, bulkDeleteItems, search, checkConflict
 *   - Segments:   listSegments, createSegment, updateSegment, deleteSegment
 *   - Folders:    searchFolders, createFolder, getFolder, updateFolder, deleteFolders, getFolderContents, searchFiles
 *   - Drive:      getOnedriveSessionStatus
 *   - Skipped:    importCsv/Excel, uploadBoardFile, uploadFile, generateAiTags (need binary fixtures — covered in test-files.ts)
 *
 * Lifecycle test creates a real board + items + fields + segments + folder, then deletes everything.
 */
import "dotenv/config"
import { ImbraceClient } from "@imbrace/sdk"

const accessToken    = process.env.IMBRACE_ACCESS_TOKEN
const organizationId = process.env.IMBRACE_ORGANIZATION_ID
const baseUrl        = process.env.IMBRACE_GATEWAY_URL ?? "https://app-gatewayv2.imbrace.co"

if (!accessToken || !organizationId) {
  console.error("Missing IMBRACE_ACCESS_TOKEN or IMBRACE_ORGANIZATION_ID")
  process.exit(1)
}

const client = new ImbraceClient({ accessToken, organizationId, baseUrl, timeout: 30_000 })
const b = client.boards

let pass = 0, fail = 0, skip = 0
const fails: string[] = []

async function step(label: string, fn: () => Promise<any>, expectFail = false) {
  process.stdout.write(`  • ${label} ... `)
  const t0 = Date.now()
  try {
    const result = await fn()
    const dt = Date.now() - t0
    const summary = JSON.stringify(result ?? {}).slice(0, 120)
    if (expectFail) { console.log(`unexpected pass [${dt}ms]: ${summary}`); fail++; fails.push(`${label} → unexpected pass`) }
    else { console.log(`✓ [${dt}ms] ${summary}`); pass++ }
  } catch (err: any) {
    const detail = err?.message ?? String(err)
    if (expectFail) { console.log(`✓ (expected fail [${detail.slice(0, 200)}])`); pass++ }
    else { console.log(`✗ ${detail.slice(0, 300)}`); fail++; fails.push(`${label} → ${detail.slice(0, 300)}`) }
  }
}

function skipped(label: string, reason: string) {
  console.log(`  - ${label}  ⏭ ${reason}`); skip++
}

function section(title: string) { console.log(`\n══ ${title} ══`) }

console.log(`\n━━━ boards (databoard) — auth: ACCESS TOKEN (npm @imbrace/sdk) ━━━`)
console.log(`gateway=${baseUrl}\norg=${organizationId}`)

// ── 1. Read-only list / get ─────────────────────────────────────────────────
section("Read-only list")
let firstBoardId: string | null = null
await step("list (limit 5)", async () => {
  const r: any = await b.list({ limit: 5 } as any)
  const data = r.data ?? r ?? []
  if (data.length > 0) firstBoardId = data[0]._id ?? data[0].id
  return { count: data.length, first: firstBoardId }
})
await step("list (sort by name asc)", () => b.list({ limit: 3, sort: "name" } as any))
await step("list (types=Contacts)", () => b.list({ limit: 3, types: "Contacts" } as any))

if (firstBoardId) {
  await step("get(firstBoardId)", () => b.get(firstBoardId!))
  await step("listItems(firstBoardId, limit=2)", () => b.listItems(firstBoardId!, { limit: 2 } as any))
  await step("listSegments(firstBoardId)", () => b.listSegments(firstBoardId!))
  await step("exportCsv(firstBoardId)", async () => {
    const csv = await b.exportCsv(firstBoardId!)
    return { length: typeof csv === "string" ? csv.length : 0 }
  })
} else {
  skipped("get / listItems / listSegments / exportCsv", "no boards in org")
}

// ── 2. Board lifecycle ──────────────────────────────────────────────────────
section("Board lifecycle (create → update → delete)")
const stamp = Date.now()
const testBoardName = `sdk-test-board-${stamp}`
let createdBoardId: string | null = null

await step("create (type=General)", async () => {
  const r: any = await b.create({
    name: testBoardName,
    description: "SDK test board — auto-cleanup",
    type: "General",
    show_id: false,
  } as any)
  createdBoardId = r._id ?? r.id
  return { id: createdBoardId, name: r.name, type: r.type }
})

if (createdBoardId) {
  await step("update (rename)", async () => {
    const r: any = await b.update(createdBoardId!, { name: `${testBoardName}-updated`, description: "updated" } as any)
    return { id: r._id ?? r.id, name: r.name }
  })

  // ── 3. Fields ─────────────────────────────────────────────────────────────
  // SDK now returns BoardField directly (was Board pre-fix #6).
  section("Fields lifecycle")
  let createdFieldId: string | null = null
  await step("createField (ShortText)", async () => {
    const r: any = await b.createField(createdBoardId!, {
      name: "test_field",
      type: "ShortText",
    } as any)
    createdFieldId = r._id ?? r.id ?? null
    return { fieldId: createdFieldId, fieldName: r.name, fieldType: r.type }
  })

  if (createdFieldId) {
    await step("updateField (rename)", async () => {
      const r: any = await b.updateField(createdBoardId!, createdFieldId!, {
        name: "test_field_renamed",
      } as any)
      return { id: r._id ?? r.id, name: r.name }
    })

    await step("deleteField", async () => {
      await b.deleteField(createdBoardId!, createdFieldId!)
      return { deleted: createdFieldId }
    })
  } else {
    skipped("updateField / deleteField", "createField did not return a field id")
  }

  // ── 4. Items ──────────────────────────────────────────────────────────────
  // Get identifier field id from board (the one with is_identifier:true).
  // Item create body uses `{fields: [{board_field_id, value}]}`,
  // update body asymmetrically uses `{fields: [], data: [{key, value}]}` — SDK drift.
  section("Items lifecycle")
  let identifierFieldId: string | null = null
  await step("get board identifier field id", async () => {
    const board: any = await b.get(createdBoardId!)
    const idField = (board.fields ?? []).find((f: any) => f.is_identifier)
    identifierFieldId = idField?._id ?? null
    return { identifierFieldId, name: idField?.name }
  })

  let createdItemId: string | null = null
  if (identifierFieldId) {
    await step("createItem", async () => {
      const r: any = await b.createItem(createdBoardId!, {
        fields: [{ board_field_id: identifierFieldId, value: "hello" }],
      } as any)
      createdItemId = r._id ?? r.id
      return { id: createdItemId }
    })
  } else {
    skipped("createItem", "no identifier field found on board")
  }

  if (createdItemId) {
    await step("getItem", async () => {
      const r: any = await b.getItem(createdBoardId!, createdItemId!)
      return { id: r._id ?? r.id }
    })

    await step("updateItem [asymmetric body: fields[] + data[] — SDK drift]", async () => {
      const r: any = await b.updateItem(createdBoardId!, createdItemId!, {
        fields: [],
        data: [{ key: identifierFieldId, value: "updated" }],
      } as any)
      return { id: r._id ?? r.id }
    })

    await step("listItems(limit=5)", () => b.listItems(createdBoardId!, { limit: 5 } as any))
    await step("search({q:'hello'})", () => b.search(createdBoardId!, { q: "hello", limit: 5 } as any))

    await step("deleteItem", async () => {
      await b.deleteItem(createdBoardId!, createdItemId!)
      return { deleted: createdItemId }
    })
  } else {
    skipped("getItem / updateItem / listItems / search / deleteItem", "createItem did not return an id")
  }

  // ── 5. Segments ───────────────────────────────────────────────────────────
  section("Segments lifecycle")
  let createdSegmentId: string | null = null
  await step("createSegment", async () => {
    const r: any = await b.createSegment(createdBoardId!, {
      name: `sdk-test-segment-${stamp}`,
      filter: {},
    } as any)
    createdSegmentId = r._id ?? r.id
    return { id: createdSegmentId, name: r.name }
  })

  if (createdSegmentId) {
    await step("updateSegment (rename)", async () => {
      const r: any = await b.updateSegment(createdBoardId!, createdSegmentId!, {
        name: `sdk-test-segment-${stamp}-updated`,
      } as any)
      return { id: r._id ?? r.id, name: r.name }
    })

    await step("listSegments", async () => {
      const r = await b.listSegments(createdBoardId!)
      return { count: Array.isArray(r) ? r.length : 0 }
    })

    await step("deleteSegment", async () => {
      await b.deleteSegment(createdBoardId!, createdSegmentId!)
      return { deleted: createdSegmentId }
    })
  } else {
    skipped("updateSegment / deleteSegment", "createSegment did not return an id")
  }

  // ── 6. Cleanup board ──────────────────────────────────────────────────────
  section("Cleanup board")
  await step("delete(board)", async () => {
    await b.delete(createdBoardId!)
    return { deleted: createdBoardId }
  })
} else {
  skipped("update / fields / items / segments / delete", "create did not return an id")
}

// ── 7. KnowledgeHub Folders ─────────────────────────────────────────────────
section("KnowledgeHub folders lifecycle")
let createdFolderId: string | null = null
await step("searchFolders", () => b.searchFolders())
await step("createFolder [SDK should default source_type='upload' + parent_folder_id='root']", async () => {
  const r: any = await b.createFolder({
    name: `sdk-test-folder-${stamp}`,
    organization_id: organizationId,
    // Backend enum: source_type must be "upload" (only valid value observed).
    // parent_folder_id must be "root" for top-level (not null).
    // SDK should default these.
    source_type: "upload",
    parent_folder_id: "root",
  } as any)
  createdFolderId = r._id ?? r.id
  return { id: createdFolderId, name: r.name }
})

if (createdFolderId) {
  await step("getFolder", async () => {
    const r: any = await b.getFolder(createdFolderId!)
    return { id: r._id ?? r.id, name: r.name }
  })
  await step("getFolder(recursive=true)", () =>
    b.getFolder(createdFolderId!, { recursive: true } as any))
  await step("getFolderContents", () => b.getFolderContents(createdFolderId!))
  await step("searchFiles(folderId)", () => b.searchFiles({ folderId: createdFolderId! } as any))
  await step("updateFolder (rename)", async () => {
    const r: any = await b.updateFolder(createdFolderId!, {
      name: `sdk-test-folder-${stamp}-updated`,
    } as any)
    return { id: r._id ?? r.id, name: r.name }
  })

  await step("deleteFolders", async () => {
    const r = await b.deleteFolders({ ids: [createdFolderId!] } as any)
    return r
  })
} else {
  skipped("getFolder / updateFolder / deleteFolders", "createFolder did not return an id")
}

// ── 8. Drive integration (best-effort) ──────────────────────────────────────
section("Drive integration (read-only)")
if ((b as any).getOnedriveSessionStatus) {
  await step("getOnedriveSessionStatus", () => (b as any).getOnedriveSessionStatus())
} else {
  skipped("getOnedriveSessionStatus", "method not exposed in this SDK build")
}

// ── 9. Skipped (need binary fixtures, covered in test-files.ts) ─────────────
section("Skipped (need binary fixtures)")
skipped("importCsv / importExcel", "needs CSV/Excel file fixture")
skipped("uploadBoardFile", "needs file fixture")
skipped("uploadFile (folder)", "needs file fixture — covered in test-files.ts")
skipped("generateAiTags", "needs file fixture")
skipped("getLinkPreview", "needs URL fixture, optional")
skipped("getRelatedItems / linkItems / unlinkItems", "needs related board fixture")

// ── Summary ─────────────────────────────────────────────────────────────────
console.log(`\n━━━ Summary ━━━`)
console.log(`pass=${pass}  fail=${fail}  skip=${skip}`)
if (fails.length) {
  console.log("Failures:")
  for (const f of fails) console.log(`  - ${f}`)
}
process.exit(fail > 0 ? 1 : 0)
